feat(core): added support for a custom url

allardy opened the pull request:
These changes allows Botpress to be served from a different url than the root one /.
For example, you could have it at http://localhost:3000/botpress/

The path is extracted from the external URL, so setting it is mandatory. Afterwards, you can access it either with process.ROOT_PATH in the backend, or with window.ROOT_PATH in the frontend (studio, admin, modules, etc).

These changes relies heavily on the `<base href>` html5 tag, so all future URL must be **relative**, not absolute. For ex, use `assets/mod` instead of `/assets/mod`

When using the root path, do not use a slash. The variable handles them. Ex: `${window.ROOT_PATH}api/v1`

## Todo

- [x] Test without externalUrl set
- [x] Test with externalUrl set with or without final slash
- [x] Test with the binary
- [ ] Test docker
- [ ] Test on a website with a real address
- [ ] Test with nginx
- [ ] Update all related documentation
- [ ] Examples, etc....

allardy committed:
fix(core): custom paths can be defined for botpress

allardy committed:
fix more links

rndlaine reviewed (commented):
I think its more intuitive to omit the last `/` in ROOT_PATH.

For instance `${window.BOT_API_PATH}` has no `/`, `${window.ROOT_PATH}` has one.

slvnperron commented on modules/channel-web/assets/examples/embedded-webchat.html:
```diff
@@ -10,13 +13,13 @@
   <script>
     window.botpressWebChat.init({
-      host: 'http://localhost:3000',
+      host: 'http://localhost:3000/',
```
webchat's init function should add the trailing `/` to make it backward compatible and also `host` normally doesn't take a protocol (http), so should probably be renamed `baseURL` in the next version

allardy replied:
I got the good balance to get everything to work by using a single source of truth (the external URL path) with those changes. When you are on the studio, the host is not set and we load the page with its relative url (assets/bla), but it works since the base href is set by the lite view. 

We could force setting the host on the studio tho.. and change it to baseURL instead. I didn't check if it may impact something else.

slvnperron commented on modules/channel-web/assets/inject.js:
```diff
@@ -23,12 +23,12 @@ window.addEventListener('message', function(payload) {
   injectDOMElement('link', 'head', { rel: 'stylesheet', href: cssHref })
   const options = encodeURIComponent(JSON.stringify({ config: config }))
-  let iframeSrc = host + '/lite/' + botId + '/?m=channel-web&v=Embedded&options=' + options
+  let iframeSrc = host + 'lite/' + botId + '/?m=channel-web&v=Embedded&options=' + options
```
we could let the `/lite/` live and replace potential double slashes, i.e. 

```js
const SanitizeUrl = url => 
  url.replace(/^(http|https):\/\/(.+)/i, (_full, protocol, rest) => protocol + "://" + rest.replace(/\/\//ig, '/'))
```

allardy replied:
Shouldn't sanitization be done only once when it is set , instead of everywhere where we load an URL ? Seems a bit like overkill to me

slvnperron commented on modules/channel-web/src/views/lite/core/api.tsx:
```diff
@@ -30,7 +30,7 @@ export default class WebchatApi {
     this.botId = botId
     this.axiosConfig = botId
-      ? { baseURL: `${window.location.origin}/api/v1/bots/${botId}/mod/channel-web` }
+      ? { baseURL: `${window.location.origin}${window.BOT_API_PATH}/mod/channel-web` }
```
so that we could use the SanitzeUrl function above everywhere and be sure we're not breaking and prevent misuse

allardy committed:
merge minor

allardy committed:
fix url reloading the page

allardy committed:
updated url for pull and benchmark

allardy committed:
added warning when external url is not set

allardy committed:
updated doc

allardy committed:
inverted rootpath logic

allardy committed:
added caching of index files when resolved

allardy commented:
Well, I've made other big changes here. I also thought it made more sense to keep the / after the root path, so i've changed the logic.

Also added a simple caching of index files instead of always reading & parsing them from the file system.
